components/region_cache_memory_engine/src/engine.rs

```diff
@@ -27,31 +28,156 @@ fn cf_to_id(cf: &str) -> usize {
     }
 }
 
+// todo: remove it
+fn readable_user_key(user_key: &[u8]) -> String {
+    let idx = user_key.len() - ENC_KEY_SEQ_LENGTH;
+    format!(
+        "{:?}{}",
+        String::from_utf8(user_key[..idx].to_vec()).unwrap(),
+        !u64::from_be_bytes(user_key[idx..idx + ENC_KEY_SEQ_LENGTH].try_into().unwrap())
+    )
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum ValueType {
+    Deletion,
+    Value,
+}
+
+// See `compare` of InternalKeyComparator, for the same user key and same
+// sequence number, ValueType::Value is less than ValueType::Deletion
+const VALUE_TYPE_FOR_SEEK: ValueType = ValueType::Value;
+const VALUE_TYPE_FOR_SEEK_FOR_PREV: ValueType = ValueType::Deletion;
+
+impl ValueType {
+    fn encode(&self) -> u8 {
+        match self {
+            ValueType::Deletion => 0,
+            ValueType::Value => 1,
+        }
+    }
+
+    fn decode(i: u8) -> Self {
+        match i {
+            0 => ValueType::Deletion,
+            1 => ValueType::Value,
+            _ => panic!("invalid"),
+        }
+    }
+}
+
+pub struct InternalKey<'a> {
+    user_key: &'a [u8],
+    v_type: ValueType,
+    sequence: u64,
+}
+
+const ENC_KEY_SEQ_LENGTH: usize = std::mem::size_of::<u64>();
+
+#[inline]
+fn decode_key<'a>(encoded_key: &'a [u8]) -> InternalKey<'a> {
+    debug_assert!(encoded_key.len() >= ENC_KEY_SEQ_LENGTH);
+    let seq_offset = encoded_key.len() - ENC_KEY_SEQ_LENGTH;
+    let num = u64::from_be_bytes(
+        encoded_key[seq_offset..seq_offset + ENC_KEY_SEQ_LENGTH]
+            .try_into()
+            .unwrap(),
+    );
+    let sequence = num >> 8;
+    let v_type = ValueType::decode((num & 0xff) as u8);
+    InternalKey {
+        user_key: &encoded_key[..seq_offset],
+        v_type,
+        sequence,
+    }
+}
+
+#[inline]
+fn encode_key_internal<T: BufMut>(
```

Comment:
may want to add a doc comment above describing the layout of the encoded key with ascii art, e.g.,:

```rust
/// Key format:
/// ```
///  key of size n.    |  value type | sequence number
/// 0 .................... n | n + 1 ... |
/// ```
```